### PR TITLE
Dynamically Fetch Study Groups Data

### PR DESCRIPTION
**Pull request by Monica-Coira**
## Description

- Fetches study groups data when the user clicks on the "Submit Availability" button and when a user creates a new group so that the Groups Page dynamically displays all of the study groups a user is matched with.
- Displays a user's study groups as soon as the user navigates to the Groups Page, without having to refresh the page.
- Implements a loading state until all study groups data is completely fetched.
- Implements error handling when fetching data from the database.
- In the next PR, I will begin working on technical challenge 1, which is that users can be matched into study groups based on a compatibility score.

## Milestones
This works towards Milestone 6, which is that users can be matched into study groups based on availability (technical challenge 2).

## Test Plan

https://github.com/user-attachments/assets/5980c765-94e7-4f43-8417-7217e0b22224

Besides testing basic functionality, these are the corner cases I tested:

- Created a user whose preferred times don't match with any existing groups to ensure that all possible groups will still be displayed, even if they don't fall during the preferred times.
- Created a user whose preferred times only match with one existing group to ensure that only the group that matches will be displayed, not all the groups that don't fall within the preferred time range.
- Created multiple new groups to ensure that they're all displayed on the Groups Page as soon as they're created without having to refresh or navigate outside of the page.
- Attempted to fetch data with incorrect body parameters to ensure that the error messages are properly displayed on the screen.

### CODE REVIEW

File: client/src/utils/MatchByAvailability.jsx [hunk 1]
```diff
@@ -236,6 +236,17 @@ const filterByPreferredTime = async (groupsByAvailability, allUsers, fetchData,
                 }
             }
         }
+        if (groupsForUser.length === 0){
+            for (const group of groupsByAvailability){
```
[the remark this answers is not in the document]
Comment (Monica-Coira) [reply client/src/utils/MatchByAvailability.jsx:240]: If there are no existing groups that fall within the user's preferred time slot, then it adds the user to all the groups that still match their availability and the classes they're taking, but just don't fall into the preferred times.

File: client/src/components/SignupForm.jsx [hunk 2]
```diff
@@ -7,6 +7,7 @@ const SignupForm = () => {
     const [message, setMessage] = useState("")
     const navigate = useNavigate();
     const {user, setUser} = useUser();
+    const [error, setError] = useState("");
```
[the remark this answers is not in the document]
Comment (Monica-Coira) [reply client/src/components/SignupForm.jsx:10]: Yes, if the user doesn't fill out all the input fields on the Sign Up form, then the error variable is used to display this message: 
https://github.com/MetaU-Capstone-Study-Group-App/Study-Match/blob/2dc6205431e31614f1cc031951057d9b91affc10/client/src/components/SignupForm.jsx#L39-L41
